python/lsst/meas/algorithms/stamps.py

```diff
@@ -113,35 +125,44 @@ def readFitsWithOptions(filename, stamp_factory, options):
     """
     # extract necessary info from metadata
     metadata = afwFits.readMetadata(filename, hdu=0)
-    f = afwFits.Fits(filename, 'r')
-    nExtensions = f.countHdus()
     nStamps = metadata["N_STAMPS"]
-    # check if a bbox was provided
-    kwargs = {}
-    if options:
-        # gen3 API
-        if "bbox" in options.keys():
-            kwargs["bbox"] = options["bbox"]
-        # gen2 API
-        elif "llcX" in options.keys():
-            llcX = options["llcX"]
-            llcY = options["llcY"]
-            width = options["width"]
-            height = options["height"]
-            bbox = Box2I(Point2I(llcX, llcY), Extent2I(width, height))
-            kwargs["bbox"] = bbox
-    stamp_parts = {}
-    # We need to be careful because nExtensions includes the primary
-    # header data unit
-    for idx in range(nExtensions-1):
-        md = afwFits.readMetadata(filename, hdu=idx+1)
-        if md['EXTNAME'] in ('IMAGE', 'VARIANCE'):
-            reader = afwImage.ImageFitsReader(filename, hdu=idx+1)
-        elif md['EXTNAME'] == 'MASK':
-            reader = afwImage.MaskFitsReader(filename, hdu=idx+1)
-        else:
-            raise ValueError(f"Unknown extension type: {md['EXTNAME']}")
-        stamp_parts.setdefault(md['EXTVER'], {})[md['EXTNAME'].lower()] = reader.read(**kwargs)
+    has_archive = metadata["HAS_ARCHIVE"]
+    if has_archive:
+        archive_ids = metadata.getArray("ARCHIVE_IDS")
+    with afwFits.Fits(filename, 'r') as f:
+        nExtensions = f.countHdus()
+        # check if a bbox was provided
+        kwargs = {}
+        if options:
+            # gen3 API
+            if "bbox" in options.keys():
+                kwargs["bbox"] = options["bbox"]
+            # gen2 API
+            elif "llcX" in options.keys():
+                llcX = options["llcX"]
+                llcY = options["llcY"]
+                width = options["width"]
+                height = options["height"]
+                bbox = Box2I(Point2I(llcX, llcY), Extent2I(width, height))
+                kwargs["bbox"] = bbox
+        stamp_parts = {}
+        # We need to be careful because nExtensions includes the primary
+        # header data unit
+        for idx in range(nExtensions-1):
+            md = afwFits.readMetadata(filename, hdu=idx+1)
+            if md['EXTNAME'] in ('IMAGE', 'VARIANCE'):
+                reader = afwImage.ImageFitsReader(filename, hdu=idx+1)
+            elif md['EXTNAME'] == 'MASK':
+                reader = afwImage.MaskFitsReader(filename, hdu=idx+1)
+            elif md['EXTNAME'] == 'ARCHIVE_INDEX':
+                f.setHdu(idx+1)
+                archive = afwTable.io.InputArchive.readFits(f)
+                continue
+            elif md['EXTTYPE'] == 'ARCHIVE_DATA':
```

Comment:
I'm not clear what this is doing.  I'd naively assume that the extension with EXTNAME = ARCHIVE_INDEX would have EXTTYPE = ARCHIVE_DATA, or is it that the archive system knows what archive data to read?

Reply:
They're two (at least) separate HDUs. `InputArchive` knows how to handle both if you just give it the one that has EXTNAME `ARCHIVE_INDEX`. This line is just to avoid raising ValueError when we encounter the other HDU.s that was written by `OutputArchive`, which will always have EXTTYPE `ARCHIVE_DATA`.

Following the discussions in [DM-28716](https://jira.lsstcorp.org/browse/DM-28716), the _EXTNAME_ for those HDUs actually contain the type of Persistable that have been stored, so I can't check against that (since we don't know a priori what that will be), hence my using EXTTYPE here.

Following the discussions in [DM-28716](https://jira.lsstcorp.org/browse/DM-28716), the EXTNAME of

Reply:
OK